### PR TITLE
Making the HonoClient implementation extensible

### PR DESCRIPTION
**Pull request by balsmn**
Allowing HonoClient implementation to be extensible so that specific custom implementations can be instantiated without replicating the whole HonoClient implementation. 

The following methods will enable custom client instances to be instantiated
```
HonoClientImpl#newRegistrationClient()
HonoClientImpl#newCredentialsClient()
```

and enables cleanup by using the following methods
```
HonoClientImpl#removeRegistrationClient() 
HonoClientImpl#removeCredentialsClient()
```
Signed-off-by: Bala Azhagappan <balasubramanian.azhagappan@bosch-si.com>

### CODE REVIEW

**Comment by balsmn** [comment]
I had a discussion about this PR with @sysexcontrol. Unfortunately I could add him as one of the reviewers to this PR.

File: client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java [hunk 1]
```diff
@@ -361,23 +367,23 @@ private void reconnect(
...
+    public final Future<MessageSender> getOrCreateTelemetrySender(final String tenantId, final String deviceId) {
 
         Objects.requireNonNull(tenantId);
         return getOrCreateSender(
                 TelemetrySenderImpl.getTargetAddress(tenantId, deviceId),
                 () -> createTelemetrySender(tenantId, deviceId));
     }
 
-    private Future<MessageSender> createTelemetrySender(
+    private final Future<MessageSender> createTelemetrySender(
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java]: private methods cannot be overridden, no need to declare them *final*

File: client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java [hunk 2]
```diff
@@ -592,7 +598,17 @@ private void reconnect(
         return result;
     }
 
-    private Future<RequestResponseClient> newCredentialsClient(final String tenantId) {
+    /**
+     * Custom implementation of {@link CredentialsClient} can be instantiated by overriding this
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java]: Please re-phrase this JavaDoc.
The first sentence should describe shortly what this method's purpose is. Subsequent paragraphs can then go into specifics and then you may add recommendations for properly overriding the method at the end.
See http://www.oracle.com/technetwork/articles/java/index-137868.html

File: client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java [hunk 3]
```diff
@@ -642,7 +658,17 @@ private void removeCredentialsClient(final String tenantId) {
         return result;
     }
 
-    private Future<RequestResponseClient> newRegistrationClient(final String tenantId) {
+    /**
+     * Custom implementation of {@link RegistrationClient} can be instantiated by overriding this
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java]: re-phrase

File: client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java [hunk 4]
```diff
@@ -642,7 +658,17 @@ private void removeCredentialsClient(final String tenantId) {
...
+     * method. Any such instance should be scoped to the given tenantId.
+     *
```
Comment (sysexcontrol) [inline client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java]: We always put a `<p>` after the introduction to separate it from the detailed explanation.

File: client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java [hunk 5]
```diff
@@ -592,7 +598,17 @@ private void reconnect(
         return result;
     }
 
-    private Future<RequestResponseClient> newCredentialsClient(final String tenantId) {
+    /**
+     * Custom implementation of {@link CredentialsClient} can be instantiated by overriding this
+     * method. Any such instance should be scoped to the given tenantId.
+     *
```
Comment (sysexcontrol) [inline client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java]: We always put a `<p>` after the introduction to separate it from the detailed explanation.

File: client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java (hunk 1 above)
Comment (balsmn) [reply client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java]: oops. thats was not supposed to be there. will correct it.

File: client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java [hunk 6]
```diff
@@ -592,7 +598,18 @@ private void reconnect(
         return result;
     }
 
-    private Future<RequestResponseClient> newCredentialsClient(final String tenantId) {
+    /**
+     * Creates a new instance of {@link CredentialsClient} scoped for the given tenant identifier.
+     * <p>
+     * Custom implementation of {@link CredentialsClient} can be instantiated by overriding this method. Any such
+     * instance should be scoped to the given tenantId.Custom extension of {@link HonoClientImpl} must invoke
```
Comment (sysexcontrol) [inline client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java]: Missing space at beginning of sentence.

Comment (balsmn) [reply client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java]: Done

**Pull request merged**